### PR TITLE
Nobody likes Wsock32

### PR DESCRIPTION
**Pull request by Badmex**
Well TD already tried this with a different approach.

So I will go ahead and explain this system. A normal ban will ban both the cid and the ckey.

This system will make you authorize yourself one time, if you stay on the same cid. That means you can go on playing like normal if you don't change your cid, if you do change your cid you will have to redo the authorization process.

The second feature is a whitelist for 2 extra cid slots. These slots will also get checked so this is for people that constantly change between the desktop and the laptop. It is an in game command that is currently grouped under the permission_group, if a person is on the whitelist they can change the cid three times.

Table on the DB:

CREATE TABLE spoof_check
(
   `id` int(11) NOT NULL AUTO_INCREMENT,
   `whitelist` int(1) NOT NULL DEFAULT '0',
   `ckey` varchar(32) NOT NULL,
   `computerid_1` varchar(32) NOT NULL,
   `computerid_2` varchar(32),
   `computerid_3` varchar(32),
   `datetime_1` datetime NOT NULL,
   `datetime_2` datetime,
   `datetime_3` datetime,
    PRIMARY KEY (ID)
)    ENGINE=InnoDB DEFAULT CHARSET=latin1;

### CODE REVIEW

**Comment by KayeArray** [comment]
And how does this stop ban evasion?

**Comment by Badmex** [comment]
It stops people from joining with wsock32 so that they will have to use the real computerid. In a ban this id will also get banned so they can't rejoin over it again and because they can't use wsock32 at all they can't come back with a different cid.

**Comment by Kokojos** [comment]
Heh I guess we could try it for a test run, it's better than reconnecting twice each time.

+0.34 votes

**Comment by KayeArray** [comment]
But how does this stop a ban evader from just joining with a different cid, ckey and ip?

**Comment by Badmex** [comment]
You can't join with a different cid. Wsock32 will always change your cid and if you take it out you have to revert towards your original computer-cid. Everytime you change your cid you will have to do the whole process again. So if you keep on changing your cid it will never stop with the authorization process and you can't join at all. If your cid is banned you can't join with that pc so you would be unable to join with a different ckey or ip. Or well you get yourself a new computer everytime.

**Comment by KayeArray** [comment]
wsock32 only changes the cid when the person changes their ip though, so if a ban evader just changes everything he can rejoin and do the authentication without anyone being the wiser

**Comment by Badmex** [comment]
No wsock is changing the cid everytime dreamseeker will start. After clicking ok on the alert you will be thrown out of the game and have to reconnect, by opening up a new dreamseeker window. So what you are telling me is not possible.

**Comment by ghost** [comment]
> wsock32 only changes the cid when the person changes their ip though, so if a ban evader just changes everything he can rejoin and do the authentication without anyone being the wiser

Wrong, the CID is randomized each time you join a server, but this will not solve anything as will still be just as easy as just creating a new account.

**Comment by Badmex** [comment]
> Wrong, the CID is randomized each time you join a server,

Yes as I was saying each time you open a new Dreamseeker window you will get a new cid.

> ,but this will not solve anything as will still be just as easy as just creating a new account.

No I don't think you understand how this system makes wsock impossible to use.
So I will explain it again

A PC got a static cid. It does not matter how often you open dreamseeker it will stay the same.
Wsock is not static every time you log into dreamseeker the dll will give you a new cid.
My System works this way:
1. You join the server you get an alert saying please rejoin one time to authorize your cid.
2. You will be forced out of the server at the same time a db entry with your cid and ckey will be created.
3.You join again if you cid is still the same, good for you. You won't get a alert the system will never bother you again.
3.5 You join again but your cid is different oh uh no problem everybody can change his PC or maybe you are really using wsock you just gotta reauthorize yourself. This means you will be forced out of the game again. The db entry will then update with the new cid that you wanna use. If you are just using your laptop that means your cid will be the same again, system won't bother you till you are changing your cid again. If you are using wsock well your cid will be different again so you will get the alert again and then will be thrown out of the game.
4. Conclusion: If you use wsock to join the game you will get an alert every time saying that you maybe changed your PC or are using wsock. Then it will force close your window and throw you out.
5. If you decide to take out wsock you will revert back towards you normal PC cid. If you are getting banned then, well you are out of luck because a ban will ban both the ckey and cid.
6. For people that constantly use different PC's I build the whitelist that will allow you too have 2 extra cid's. So if you join with your laptop instead of your PC you will activate the second cid slot and then well have fun playing.

**Comment by Earthykiller127** [comment]
oh good, now I can finally put this thing to good use
http://puu.sh/qAWrw/f9f1d43036.jpg

**Comment by ghost** [comment]
@Earthykiller127 yep, this entire PR will solve nothing and only add more annoying systems.

**Comment by Earthykiller127** [comment]
@Chnkr If you're gonna implement something that stops griefer and it also affects legit users, then the griefers have won

**Comment by Badmex** [comment]
Good for you guys that you like to complain that much. 
This pr does what it is supposed to do and that is keep players out with the dll.
It only affects legit users one time, in the way that they will have to reconnect. Afterwards as long as they stay on the same pc it will not affect them anymore.

**Comment by ghost** [comment]
If this only affects legit users once and allows for multiple CIDs which is good.
I'll speak to Nalar about the possibility of making a sub forum for applying for more CIDs.

Is there functionality for wiping the CIDs at the moment? Just for certain cases.
If so, I'll happily +1 this, after we get the subforum up and running.

**Comment by Badmex** [comment]
Yes as gm+ you get a new button under special verbs.
This button allows you to reset ckeys, whitelist ckeys, dewhitelist ckeys.
If a person is using all three cid's and they still wanna play with a different PC the first saved cid will be changed and they need to reauthorize it.
So basically you can always play as long as you are not using wsock32, even if you are changing pc's you will just need to authorize yourself again with a second login.

**Comment by Earthykiller127** [comment]
yea, I don't even use wsock32 anymore

**Comment by SplergeBucket** [comment]
ok cool eartho
i'd like to give this a trial
should be good

File: code/modules/admin/verbs/randomverbs.dm [hunk 1]
```diff
@@ -1066,3 +1066,99 @@ var/list/datum/outfit/custom_outfits = list() //Admin created outfits
...
+
+/client/proc/whitelist_cid()
+	set name = "Whitelist-Ckey"
+	set category = "Special Verbs"
+	set desc = "Whitelist Players or delete connected cids"
+
+	if(!holder)
+		return
+
+	var/confirm = alert(src, "Choose:", "Confirm", "Whitelist/De-Whitelist", "Reset", "Cancle")
```
Comment (JohnGinnane) [inline code/modules/admin/verbs/randomverbs.dm]: "Cancle"

File: code/modules/client/client procs.dm [hunk 2]
```diff
@@ -431,3 +431,93 @@ var/next_external_rsc = 0
...
+			query_check_whiteextra.Execute()
+
+			if(query_check_whiteextra.RowCount() != 0)//One of the three slots had the same cid
+				cid_check = 1
+			else
+				var/DBQuery/query_check_white2 = dbcon.NewQuery("SELECT whitelist FROM [format_table_name("spoof_check")] WHERE ckey = '[sql_ckey]' and whitelist = '1'")
+				query_check_white2.Execute()
+
+				if(query_check_white2.RowCount() != 0) // another check.. Well kinda not needed but maybe maybe
+					var/DBQuery/query_check_cid1r = dbcon.NewQuery("SELECT ckey FROM [format_table_name("spoof_check")] WHERE ckey = '[sql_ckey]' and computerid_1 = 0")
```
Comment (JohnGinnane) [inline code/modules/client/client procs.dm]: Should check for '0' instead?

File: code/modules/mob/new_player/new_player.dm [hunk 3]
```diff
@@ -105,6 +109,10 @@
 
 	if(href_list["observe"])
 
+		if(usr.client.cid_check == 0)
+			src << "<span class='warning'>You are not authorized yet. Please rejoin the game once.</span>"
+			return 1
```
Comment (JohnGinnane) [inline code/modules/mob/new_player/new_player.dm]: Why return 1 here?

File: code/modules/mob/new_player/new_player.dm [hunk 4]
```diff
@@ -130,6 +138,10 @@
 			return 1
 
 	if(href_list["late_join"])
+		if(usr.client.cid_check == 0)
+			src << "<span class='warning'>You are not authorized yet. Please rejoin the game once.</span>"
+			return 1
```
Comment (JohnGinnane) [inline code/modules/mob/new_player/new_player.dm]: And here

File: code/modules/admin/secrets.dm [hunk 5]
```diff
@@ -146,6 +147,18 @@
...
+			var/dat = "<B>Current ckeys that are whitelisted:</B><HR>"
+			establish_db_connection()
+			if (!dbcon.IsConnected())
+				return
+
+			var/DBQuery/query_check_ckey = dbcon.NewQuery("SELECT `ckey`, `computerid_1`, `computerid_2`, `computerid_3`, `datetime_1`, `datetime_2`, `datetime_3` FROM [format_table_name("spoof_check")] WHERE whitelist = '1'")
+			if(query_check_ckey.Execute())
+				while(query_check_ckey.NextRow())
+					dat += "ckey:&ensp;[query_check_ckey.item[1]]<br>&emsp;cid_1:&ensp;[query_check_ckey.item[2]]&emsp;date:&ensp;[query_check_ckey.item[5]]<br>&emsp;cid_2:&ensp;[query_check_ckey.item[3]]&emsp;date:&ensp;[query_check_ckey.item[6]]<br>&emsp;cid_3:&ensp;[query_check_ckey.item[4]]&emsp;date:&ensp;[query_check_ckey.item[7]]<br>"
+				usr << browse(dat, "window=showadmins;size=600x500")
```
Comment (Carbonhell) [inline code/modules/admin/secrets.dm]: is the size enough?have you checked?

File: code/modules/admin/verbs/randomverbs.dm [hunk 6]
```diff
@@ -1066,3 +1066,99 @@ var/list/datum/outfit/custom_outfits = list() //Admin created outfits
...
+			if(!input)
+				return
+
+			establish_db_connection()
+			if (!dbcon.IsConnected())
+				return
+
+			var/sql_ckey = sanitizeSQL(input)
+
+			var/DBQuery/query_check_ckey = dbcon.NewQuery("SELECT ckey FROM [format_table_name("spoof_check")] WHERE ckey = '[sql_ckey]'")
```
Comment (Carbonhell) [inline code/modules/admin/verbs/randomverbs.dm:1101]: fix

File: code/modules/client/client procs.dm [hunk 7]
```diff
@@ -431,3 +431,93 @@ var/next_external_rsc = 0
 
 		//Precache the client with all other assets slowly, so as to not block other browse() calls
 		getFilesSlow(src, asset_cache, register_asset = FALSE)
+
+client/New()
```
Comment (Carbonhell) [inline code/modules/client/client procs.dm]: starting slash

Comment (Carbonhell) [inline code/modules/client/client procs.dm]: and btw the new proc is already at line 109,merge this stuff there

Comment (Carbonhell) [inline code/modules/client/client procs.dm]: or even better make this a different proc and call it inside client New()?

File: code/modules/client/client procs.dm [hunk 8]
```diff
@@ -431,3 +431,93 @@ var/next_external_rsc = 0
...
+	. = ..()
+	establish_db_connection()
+	if (!dbcon.IsConnected())
+		return
+
+	var/sql_ckey = sanitizeSQL(src.ckey)
```
Comment (Carbonhell) [inline code/modules/client/client procs.dm]: src. aaaAAAAAAAAAAAA

**Comment by Lawligagger** [comment]
carbon pls

**Comment by Carbonhell** [comment]
Code seems okay

**Comment by Lawligagger** [comment]
mergaroni

**Pull request merged**